### ISSUE TITLE
Suggestion for cache layer at CoWIN API side

### ISSUE THREAD

**Issue by prathamesh-gharat** [cowinapi/developer.cowin#401]
I came across someone's Tweet saying that CoWIN is now also caching the authenticated slot availability APIs, which suggests that CoWIN must be handling many API calls and facing performance issues.

Instead of having a fixed cache expiry time why not implement stale-while-revalidate at the cache layer?

Example Nginx config if you are hosting a public proxy of the CoWIN APIs with 3 seconds cache at your end.
Cowin could replicate the same at their end if they aren't already using a stale-while-revalidate implementation.
This would also address the 'Thundering Herd' problem if they are facing this issue.
```
http {
proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=cowin:1m max_size=512m inactive=60m;
proxy_temp_path /var/cache/nginx/cowin_proxy_temp;
}

server {
    listen 8080;
    server_name cowin-cache-proxy.test *.cowin-cache-proxy.test;

    location / {
        # Handle SNI
        proxy_ssl_server_name on;

        # Proxy all requests to CoWIN APIs
        proxy_pass https://cdn-api.co-vin.in/;

        # Proxy cache config
        proxy_cache cowin;
        proxy_cache_valid  200 3s;
        proxy_cache_valid  404 10s;
        proxy_cache_use_stale updating;
        proxy_cache_background_update on;
        proxy_cache_key $scheme$proxy_host$uri$is_args$args;
        proxy_ignore_headers Cache-Control;
        proxy_ignore_headers Set-Cookie;
        proxy_cache_revalidate on;

        proxy_hide_header Age;
        proxy_hide_header ETag;
        add_header X-Proxy-Cache $upstream_cache_status;
        add_header Cache-Control "max-age=0, must-revalidate";
    }

}
```

If there is a way to contact someone and get a certain IP whitelisted for the limit per IP restrictions please let me know.

**Comment by prathamesh-gharat** [comment]
Apache Bench test.

10000 requests with 10 users (concurrency).

**Command**
```
ab -c 10 -n 10000 "http://cowin-cache-proxy.test:8080/api/v2/appointment/sessions/publ
ic/calendarByDistrict?district_id=395&date=06-06-2021"
```

Output
```
This is ApacheBench, Version 2.3
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking cowin-cache-proxy.test (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx/1.16.0
Server Hostname:        cowin-cache-proxy.test
Server Port:            8080

Document Path:          /api/v2/appointment/sessions/public/calendarByDistrict?district_id=395&date=06-06-2021
Document Length:        188587 bytes

Concurrency Level:      10
Time taken for tests:   5.923 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      1897250454 bytes
HTML transferred:       1885870000 bytes
Requests per second:    1688.31 [#/sec] (mean)
Time per request:       5.923 [ms] (mean)
Time per request:       0.592 [ms] (mean, across all concurrent requests)
Transfer rate:          312806.62 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       2
Processing:     3    6   1.0      5      44
Waiting:        1    5   0.8      5       9
Total:          3    6   1.0      6      46

Percentage of the requests served within a certain time (ms)
  50%      6
  66%      6
  75%      6
  80%      7
  90%      7
  95%      8
  98%      8
  99%      9
 100%     46 (longest request)
```

Key points in output
```
Complete requests:      10000 ### No failures
Requests per second:    1688.31 [#/sec] (mean) ### It's fast?

Percentage of the requests served within a certain time (ms)
99%      9  ### 99% of the requests were cached (denoted/assumed by 9ms response time)
```

**Comment by Ashik-1997** [comment]
Calling public API within the limits (3 per minute) , server located in India(GCP Mumbai) and user-agent header is added. Still getting 403 sometimes

**Comment by safiyat** [comment]
Let us not talk about public API access. We know it is useless if one wants to query for appointments.

But as the OP said, it looks like even the private APIs are returning cached data.

Upon logging in, it shows that slots are available. But when we go to book, it complaints that they are full.
But when we go back to the previous page and make a fresh query for availability, it still shows slots are available. (I have turned off caching on my browser.)

So, apparently even private API responses are cached now.

Also, no idea how, but even when private API queries are spaced by 4 seconds or more, the server recognizes that it is a bot and denies the request with a 403 response: `{"message":"User is not authorized to access this resource with an explicit deny"}`.

Any ideas, anyone?

**Comment by prathamesh-gharat** [comment]
Hmm... their servers must be handing an immense volume of requests for them to be caching anything. At that point they think at least returning a cached response is better than our servers crashing.

If they are using a single DB then things might get slow as traffic increases if they are not scaling horizontally.
If the problem is at the database server side then they could use this one
https://www.cockroachlabs.com/ They could use DB which scales horizontally.

**Comment by udayogra** [comment]
Protected APIs are cached for 5 mins per pincode search. There is a way to bypass it and I have explained it here 👍 
https://www.youtube.com/watch?v=hf44_-qqddY

By the way public API is not at all cached. I am about to find slots instantly

**Comment by nonrealjunco** [comment]
> Protected APIs are cached for 5 mins per pincode search. There is a way to bypass it and I have explained it here 👍
> https://www.youtube.com/watch?v=hf44_-qqddY
> 
> By the way public API is not at all cached. I am about to find slots instantly

 **jor jor se chilla ke sabko scheme bata de.jpg**

**Comment by prathamesh-gharat** [comment]
The entire point of this thread was that CoWIN could use a better cache strategy.
The 5 minutes cache feels too long from my perspective.
"Bypassing" the cache further creates a divide between who is able to get slots when there is a shortage of slots.

**Comment by udayogra** [comment]
> The entire point of this thread was that CoWIN could use a better cache strategy.
> The 5 minutes cache feels too long from my perspective.
> "Bypassing" the cache further creates a divide between who is able to get slots when there is a shortage of slots.

Dont worry, they have changed it. 5 mins has been reduced to 1 minute

**Comment by nonrealjunco** [comment]
> > The entire point of this thread was that CoWIN could use a better cache strategy.
> > The 5 minutes cache feels too long from my perspective.
> > "Bypassing" the cache further creates a divide between who is able to get slots when there is a shortage of slots.
> 
> Dont worry, they have changed it. 5 mins has been reduced to 1 minute

Nope. Its still 5mins. Once you hit the API they record your IP I think. And then the AGE header gets updated in seconds. Even If you hit after 7 mins it will show AGE as 120 and will not start from 0.

**Comment by udayogra** [comment]
I checked again. Indeed it has been increased to 5 mins again. Yesterday whole day it was 1 minute

**Comment by udayogra** [comment]
So this is my finding :

sessions/public/findByDistrict is realtime and not cached
sessions/public/calendarByDistrict is cached for 5 mins
sessions/calendarByDistrict is cached for 1 minute // this is what cowin portal invokes
